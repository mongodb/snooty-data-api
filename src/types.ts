export interface ResponseAsset {
  filenames: string[];
  data: BinaryData;
}
export type ResponseAssets = Record<string, ResponseAsset>
