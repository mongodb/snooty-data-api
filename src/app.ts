import express from 'express';
import dotenv from 'dotenv';
import { MongoClient } from 'mongodb';
// Configure dotenv early so env variables can be read in imported files
dotenv.config();
import buildsRouter from './routes/builds';
import { setupClient } from './services/database';

interface AppSettings {
  mongoClient?: MongoClient;
};

export const setupApp = async ({ mongoClient }: AppSettings) => {
  if (mongoClient) {
    await setupClient(mongoClient);
  }

  const app = express();
  app.use('/builds', buildsRouter);

  return app;
};
