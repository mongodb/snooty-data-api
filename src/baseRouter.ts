import express from 'express';
import buildsRouter from './routes/builds';
import projectsRouter from './routes/projects';
import userRouter from './routes/user';

// This router will serve as a wrapper for all actual endpoints. This is to allow
// a separate /pre-prod prefix to exist for the same endpoints for testing purposes.
const router = express.Router();

router.use('/builds', buildsRouter);
router.use('/projects', projectsRouter);
router.use('/user', userRouter);
router.use('/prod/builds', buildsRouter);
router.use('/prod/projects', projectsRouter);

export default router;
