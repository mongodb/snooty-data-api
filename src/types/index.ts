export interface DataStreamOptions {
  reqId?: string;
  reqTimestamp?: number;
  updatedAssetsOnly?: boolean;
}
