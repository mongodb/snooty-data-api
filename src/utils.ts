import { Request } from 'express';

/**
 * Checks for req-id Request Header. Returns an empty string if the header is not
 * a truthy string.
 *
 * @param req
 * @returns
 */
export const getRequestId = (req: Request) => {
  const reqId = req.headers['req-id'];
  if (!reqId) {
    return undefined;
  } else if (Array.isArray(reqId)) {
    return undefined;
  } else {
    return reqId;
  }
};
