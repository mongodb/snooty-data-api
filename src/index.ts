import dotenv from 'dotenv';
// Configure dotenv early so env variables can be read in imported files
dotenv.config();
import { setupApp } from './app';
import { closeDBConnection } from './services/database';
import { initiateLogger } from './services/logger';

const PORT = process.env.PORT || 3000;
const logger = initiateLogger();

const startServer = async () => {
  const app = await setupApp({});

  const server = app.listen(PORT, () => {
    logger.info(`Server listening on port: ${PORT}`);
  });

  process.on('SIGINT', async () => {
    logger.info('SIGINT signal received');
    await closeDBConnection();
    server.close();
  });
};

try {
  startServer();
} catch (e) {
  logger.error(`Fatal error: ${e}`);
  process.exit(1);
}
