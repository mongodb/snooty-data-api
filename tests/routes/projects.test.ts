import { MongoClient } from 'mongodb';
import request from 'supertest';
import { setupApp } from '../../src/app';
import { sampleMetadata } from '../sampleData/metadata';

const timestamp = 1685714694420;

describe('Test projects routes', () => {
  // process.env.ATLAS_URI should be defined by default in globalSetup.ts
  const client = new MongoClient(process.env.ATLAS_URI!);
  let app: Express.Application;

  beforeAll(async () => {
    Date.now = jest.fn(() => timestamp);
    app = await setupApp({ mongoClient: client });
  });

  afterAll(async () => {
    await client.close();
  });

  it('should return all data based on project ID', async () => {
    const res = await request(app).get('/projects/docs/master/documents');
    expect(res.status).toBe(200);
    const data = res.text.split('\n');
    expect(data).toMatchSnapshot();
  });

  it('should return documents updated after given timestamp', async () => {
    const prevBuildTime = sampleMetadata[0].created_at;
    const timestamp = new Date(prevBuildTime).getTime();
    const res = await request(app).get(`/projects/docs/master/documents/updated/${timestamp}`);
    expect(res.status).toBe(200);
    const data = res.text.split('\n');
    expect(data).toMatchSnapshot();
  });
});
