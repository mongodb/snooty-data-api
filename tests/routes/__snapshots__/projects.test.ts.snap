// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test projects routes should return all data based on project ID 1`] = `
[
  "{"type":"timestamp","data":1685714694420}",
  "{"type":"metadata","data":{"_id":"646bae50c41d7b9a472b6078","project":"docs","branch":"master","title":"MongoDB Manual","eol":false,"slugToTitle":{},"toctreeorder":{},"parentPaths":{},"static_files":{},"build_id":"642ec854c38bedd45ed31234","created_at":"2023-04-06T14:25:40.000Z"}}",
  "{"type":"page","data":{"_id":"646fa1958c4f35b1d3d83edf","page_id":"docs/docsworker-xlarge/master/data-center-awareness","filename":"data-center-awareness.txt","ast":{"foo":"no assets"},"static_assets":[],"created_at":"2023-04-06T13:25:40.000Z","updated_at":"2023-04-06T13:25:40.000Z","deleted":false}}",
  "{"type":"page","data":{"_id":"646fa1958c4f35b1d3d83ee0","page_id":"docs/docsworker-xlarge/master/data-center-foo","filename":"data-center-foo.txt","ast":{"foo":"deleted"},"static_assets":[],"created_at":"2023-04-06T13:25:40.000Z","updated_at":"2023-04-06T14:25:40.000Z","deleted":true}}",
  "{"type":"page","data":{"_id":"646fa1958c4f35b1d3d83ee1","page_id":"docs/docsworker-xlarge/master/tutorial/insert-documents","filename":"tutorial/insert-documents.txt","ast":{"foo":"1 asset example"},"static_assets":[{"checksum":"d4dbe419766d19842ba5624cb41f6b873a513b620049e759a923894bef17742c","key":"/images/CSFLE_Master_Key_KMS.png"}],"created_at":"2023-04-06T13:25:40.000Z","updated_at":"2023-04-06T14:25:40.000Z","deleted":false}}",
  "{"type":"page","data":{"_id":"646fa1958c4f35b1d3d83ee2","page_id":"docs/docsworker-xlarge/master/tutorial/update-documents","filename":"tutorial/update-documents.txt","ast":{},"static_assets":[{"checksum":"ba1b4507fc789827edd02f979130e9633a0020b498cba915c0773403b37aedd5","key":"/images/compass-update-doc-button.png"},{"checksum":"9617b2e1a8065e4a1699c15fc8ea1e085a41b4021debeaef1793a6cb431a9659","key":"/images/compass-update-edit-mode.png"}],"created_at":"2023-04-06T13:25:40.000Z","updated_at":"2023-04-06T13:25:40.000Z","deleted":false}}",
  "{"type":"asset","data":{"checksum":"9617b2e1a8065e4a1699c15fc8ea1e085a41b4021debeaef1793a6cb431a9659","assetData":{"$binary":{"base64":"YXNzZXQtZXhhbXBsZQ==","subType":"00"}},"filenames":["/images/compass-update-edit-mode.png"]}}",
  "{"type":"asset","data":{"checksum":"ba1b4507fc789827edd02f979130e9633a0020b498cba915c0773403b37aedd5","assetData":{"$binary":{"base64":"YXNzZXQtZXhhbXBsZQ==","subType":"00"}},"filenames":["/images/compass-update-doc-button.png"]}}",
  "{"type":"asset","data":{"checksum":"d4dbe419766d19842ba5624cb41f6b873a513b620049e759a923894bef17742c","assetData":{"$binary":{"base64":"YXNzZXQtZXhhbXBsZQ==","subType":"00"}},"filenames":["/images/CSFLE_Master_Key_KMS.png"]}}",
]
`;

exports[`Test projects routes should return documents updated after given timestamp 1`] = `
[
  "{"type":"timestamp","data":1685714694420}",
  "{"type":"metadata","data":{}}",
  "{"type":"page","data":{"_id":"646fa1958c4f35b1d3d83ee0","page_id":"docs/docsworker-xlarge/master/data-center-foo","filename":"data-center-foo.txt","ast":{"foo":"deleted"},"static_assets":[],"created_at":"2023-04-06T13:25:40.000Z","updated_at":"2023-04-06T14:25:40.000Z","deleted":true}}",
  "{"type":"page","data":{"_id":"646fa1958c4f35b1d3d83ee1","page_id":"docs/docsworker-xlarge/master/tutorial/insert-documents","filename":"tutorial/insert-documents.txt","ast":{"foo":"1 asset example"},"static_assets":[{"checksum":"d4dbe419766d19842ba5624cb41f6b873a513b620049e759a923894bef17742c","key":"/images/CSFLE_Master_Key_KMS.png"}],"created_at":"2023-04-06T13:25:40.000Z","updated_at":"2023-04-06T14:25:40.000Z","deleted":false}}",
  "{"type":"asset","data":{"checksum":"d4dbe419766d19842ba5624cb41f6b873a513b620049e759a923894bef17742c","assetData":{"$binary":{"base64":"YXNzZXQtZXhhbXBsZQ==","subType":"00"}},"filenames":["/images/CSFLE_Master_Key_KMS.png"]}}",
]
`;
